pkg/storage/stores/shipper/compactor/deletion/request_handler.go

```diff
@@ -39,21 +39,21 @@ func (dm *DeleteRequestHandler) AddDeleteRequestHandler(w http.ResponseWriter, r
 	ctx := r.Context()
 	userID, err := tenant.TenantID(ctx)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
+		serverutil.JSONError(w, http.StatusBadRequest, err.Error())
```

Comment:
I'm not too familiar with how HTTP errors are handled, but are they not all handled `pkg/util/server/error.go -> WriteError`? If we use `http.Error` here does it kill the request and write this error directly?

Reply:
`WriteError` handles generic errors as `Internal Server Error`. Therefore we bypass `WriteError` and call `JSONError` directly, so we can specify a status code and an error message.

`http.Error` sets the status code, content type header (`text/plain`) and writes the response body. Usually you want to return thereafter.

Reply:
OK thanks for that. I was looking for a way that we could somehow add middleware to catch these `http.Error` calls, but they write directly to the `http.ResponseWriter` :(

I would've loved to have stuck with idiomatic gorilla/mux & net/http, but it seems we don't have a good layer of indirection here.

What you've got here looks like a good solution :+1:

Reply:
An option to catch and resolve errors in a middleware would be, to write an occurring error to the request context and check for it in the middleware handler. I don't think that is very idiomatic, though. Writing the error to the response writer seems to be the common way.